text/3606-temporary-lifetimes-in-tail-expressions.md

```diff
@@ -0,0 +1,163 @@
+# Temporary lifetimes in tail expressions
+
+- Feature Name: `tail_temporaries_shorter_lifetime`
+- Start Date: 2023-05-04
+- RFC PR: [rust-lang/rfcs#3606](https://github.com/rust-lang/rfcs/pull/3606)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+In the next edition, drop temporaries in tail expressions *before* dropping locals, rather than after.
+
+![before-after](https://hackmd.io/_uploads/HyVB0FtkA.svg)
+
+# Motivation
+
+Temporaries in the tail expression in a block live longer than the block itself,
+so that e.g. `{expr;}` and `{expr}` can behave very differently.
+
+For example, this fails to compile:
+
+```rust
+// This fails to compile!
+fn f() -> usize {
+    let c = RefCell::new("..");
+    c.borrow().len() // ERROR!!!
+}
+```
+
+The temporary `std::cell::Ref` created in the tail expression will be dropped
+after the local `RefCell` is dropped, resulting in a lifetime error.
+
+This leads to having to add seemingly unnecessary extra `let` statements
+or having to add seemingly unnecessary semicolons:
+
+```rust
+fn main() {
+    let c = std::cell::RefCell::new(123);
+
+    if let Ok(mut b) = c.try_borrow_mut() {
+        *b = 321;
+    }; // <-- Error if you remove the semicolon!
+}
+```
+
+Both of these examples will compile fine after the proposed change.
+
+# Guide-level explanation
+
+Temporaries are normally dropped at the end of the statement.
+
+The tail expression of a block
+(such as a function body, if/else body, match arm, block expression, etc.)
+is not a statement, so has its own rule:
+
+- Starting in Rust 2024,
+  temporaries in tail expressions are dropped after evaluating the tail expression,
+  but before dropping any local variables of the block.
+
+For example:
+
+```rust
+fn f() -> usize {
+    let c = RefCell::new("..");
+    c.borrow().len() // Ok in Rust 2024
+}
+```
+
+The `.borrow()` method returns a (temporary) `Ref` object that borrows `c`.
+Starting in Rust 2024, this will compile fine,
+because the temporary `Ref` is dropped before dropping local variable `c`.
+
+# Reference-level explanation
+
+For blocks/bodies/arms whose `{}` tokens come from Rust 2024 code,
+temporaries in the tail expression will be dropped *before* the locals of the block are dropped.
```

Comment:
iiuc does this imply these drop orders in the following compositions?

<table><tr><th>Code</th><th>Drop order?</th></tr>
<tr><td>

```rust
edition_2021! {
    let a = ..;
    edition_2021! {
        let b = ..;
        temp_1().result
    }
}
```

</td><td>

Current situation

1. b
2. a
3. temp_1()

</td></tr><tr><td>

```rust
edition_2024! {
    let c = ..;
    edition_2024! {
        let d = ..;
        temp_2().result
    }
}
```

</td><td>

This RFC

1. temp_2()
2. d
3. c

</td></tr><tr><td>

```rust
edition_2021! {
    let e = ..;
    edition_2024! {
        let f = ..;
        temp_3().result
    }
}
```

</td><td>

'24 block in '21 function

1. temp_3()
2. f
3. e

</td></tr><tr><td>

```rust
edition_2024! {
    let g = ..;
    edition_2021! {
        let h = ..;
        temp_4().result
    }
}
```

</td><td>

'21 block in '24 function (???)

1. h
2. temp_4()
3. g


</td></tr></table>

Reply:
Correct!

I don't expect to see people mixing editions like that though ^^'. The point is that a block expanded from a macro will behave the way the macro author expected. The edition of whoever wrote the block is used.